async_producer.go

```diff
@@ -217,8 +215,15 @@ func (p *asyncProducer) topicDispatcher() {
 		}
 
 		if msg.flags&shutdown != 0 {
-			Logger.Println("Producer shutting down.")
-			break
+			shuttingDown = true
+			continue
+		} else if msg.retries == 0 {
+			if shuttingDown {
+				p.returnError(msg, ErrShuttingDown)
```

Comment:
This reduces the inflight counter, but it was never incremented for this message. We should probably move the `p.inFlight.Add(1)` up so it always gets executed.